protocol-units/da/m1/util/src/ir_blob.rs

```diff
@@ -0,0 +1,208 @@
+		AffinePoint, CurveArithmetic, FieldBytesSize, PrimeCurve, Scalar,
+	},
+	hazmat::{DigestPrimitive, SignPrimitive, VerifyPrimitive},
+	signature::{digest::Digest, DigestVerifier},
+	SignatureSize, SigningKey, VerifyingKey,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct InnerSignedBlobV1Data {
+	pub blob: Vec<u8>,
+	pub timestamp: u64,
+}
+
+impl InnerSignedBlobV1Data {
+	pub fn new(blob: Vec<u8>, timestamp: u64) -> Self {
+		Self { blob, timestamp }
+	}
+
+	pub fn try_to_sign<C>(
+		self,
+		signing_key: &SigningKey<C>,
+	) -> Result<InnerSignedBlobV1, anyhow::Error>
+	where
+		C: PrimeCurve + CurveArithmetic + DigestPrimitive + PointCompression,
+		Scalar<C>: Invert<Output = CtOption<Scalar<C>>> + SignPrimitive<C>,
+		SignatureSize<C>: ArrayLength<u8>,
+		AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C> + VerifyPrimitive<C>,
+		FieldBytesSize<C>: ModulusSize,
+	{
+		let mut hasher = C::Digest::new();
+		hasher.update(self.blob.as_slice());
+		hasher.update(&self.timestamp.to_be_bytes());
+		let prehash = hasher.finalize();
+		let prehash_bytes = prehash.as_slice();
+
+		let (signature, _recovery_id) = signing_key.sign_prehash_recoverable(prehash_bytes)?;
+
+		Ok(InnerSignedBlobV1 {
+			data: self,
+			signature: signature.to_vec(),
+			signer: signing_key.verifying_key().to_sec1_bytes().to_vec(),
+			id: prehash_bytes.to_vec(),
+		})
+	}
+
+	pub fn try_verify<C>(&self, signature: &[u8], signer: &[u8]) -> Result<(), anyhow::Error>
+	where
+		C: PrimeCurve + CurveArithmetic + DigestPrimitive + PointCompression,
+		Scalar<C>: Invert<Output = CtOption<Scalar<C>>> + SignPrimitive<C>,
+		SignatureSize<C>: ArrayLength<u8>,
+		AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C> + VerifyPrimitive<C>,
+		FieldBytesSize<C>: ModulusSize,
+	{
+		let mut hasher = C::Digest::new();
+		hasher.update(self.blob.as_slice());
+		hasher.update(&self.timestamp.to_be_bytes());
+
+		let verifying_key = VerifyingKey::<C>::from_sec1_bytes(signer)?;
+		let signature = ecdsa::Signature::from_bytes(signature.into())?;
+
+		match verifying_key.verify_digest(hasher, &signature) {
+			Ok(_) => Ok(()),
+			Err(_) => Err(anyhow::anyhow!("Failed to verify signature")),
+		}
+	}
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct InnerSignedBlobV1 {
+	pub data: InnerSignedBlobV1Data,
+	pub signature: Vec<u8>,
+	pub signer: Vec<u8>,
+	pub id: Vec<u8>,
+}
+
+impl InnerSignedBlobV1 {
+	pub fn try_verify<C>(&self) -> Result<(), anyhow::Error>
+	where
+		C: PrimeCurve + CurveArithmetic + DigestPrimitive + PointCompression,
+		Scalar<C>: Invert<Output = CtOption<Scalar<C>>> + SignPrimitive<C>,
+		SignatureSize<C>: ArrayLength<u8>,
+		AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C> + VerifyPrimitive<C>,
+		FieldBytesSize<C>: ModulusSize,
+	{
+		self.data.try_verify::<C>(self.signature.as_slice(), self.signer.as_slice())
+	}
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
```

Comment:
How do `Serialize` and `Deserialize` represent the enum? Should we add some attributes there?